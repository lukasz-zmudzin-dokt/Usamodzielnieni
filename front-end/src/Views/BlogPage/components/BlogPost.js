// placeholder taki

import React from "react";

const BlogPost = data => (
  <>
    <p>Kategoria: {data.category}</p>
    <p>content: {data.content}</p>
  </>
);

export default BlogPost;
