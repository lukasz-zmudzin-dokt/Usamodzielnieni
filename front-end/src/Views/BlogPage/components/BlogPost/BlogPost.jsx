import React from "react";
import { Row, Badge, Col } from "react-bootstrap";
import { IndexLinkContainer } from "react-router-bootstrap";

const stripFromTags = str => {
  console.log(str);
  str = str.replace(/(<([^>]+)>)|(<([^>]+)\.\.\.)/gi, '');
  return str;
};

const BlogPost = ({data, ...rest}) => {
  const str = stripFromTags(data.summary);
  return (
    <IndexLinkContainer to={`/blog/blogpost/${data.id}`}>
      <Row {...rest}>
        <Col>
          <h3>{data.category}</h3>
          <p>{`${str.substring(0, 100)}...`}</p>
          <Row>
            <Col>
              <small>Tagi:</small>
              {data.tags.map((tag, i) => (
                <Badge variant="info" key={i} className="ml-1">
                  {tag}
                </Badge>
              ))}
            </Col>
          </Row>
        </Col>
      </Row>
    </IndexLinkContainer>
  );
};

export default BlogPost;
