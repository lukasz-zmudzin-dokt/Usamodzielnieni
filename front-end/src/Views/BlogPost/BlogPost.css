.blogpost_comment_card {
    margin-top: 1rem;
}

@media (max-width: 768px){
    .blogpost_container {
        width: 100%;
    }
}

@media (orientation: landscape) {
    .blogpost_container {
        width: 60vw;
    }
}