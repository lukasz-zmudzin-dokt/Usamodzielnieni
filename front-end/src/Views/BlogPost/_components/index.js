import CommentForm from "./CommentForm/CommentForm";
import CommentsList from "./CommentsList/CommentsList";
import CommentItem from "./CommentItem/CommentItem";
import BlogContent from "./BlogContent/BlogContent";

export { CommentForm, CommentsList, CommentItem, BlogContent };
