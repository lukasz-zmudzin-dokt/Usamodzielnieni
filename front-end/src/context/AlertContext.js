import React, { useState, createContext } from "react";

export const AlertContext = createContext({
  open: false,
  message: "",
  setOpen: () => {},
});

export const AlertProvider = (props) => {
  const [open, setOpen] = useState(false);
  const [message, setMessage] = useState("");
  const [variant, setVariant] = useState("danger") // success
  const data = {
    open,
    message,
    variant,
    changeVisibility: (newOpen) => setOpen(newOpen),
    changeMessage: (newMessage,newVariant = variant) => {
      setOpen(false);
      setMessage(newMessage);
      setVariant(newVariant);
    },
    showAlert:  (newMessage,newVariant) => {
      data.changeMessage(newMessage,newVariant);
      data.changeVisibility(true);
    }
  };
  return <AlertContext.Provider value={data} {...props} />;
};
