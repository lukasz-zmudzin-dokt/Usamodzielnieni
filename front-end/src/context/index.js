export { UserContext, UserProvider } from "context/UserContext";
export { AlertContext, AlertProvider } from "context/AlertContext";
export { ChatContext, ChatProvider } from "context/ChatContext";
export {
  NotificationsContext,
  NotificationsProvider,
} from "context/NotificationsContext";
