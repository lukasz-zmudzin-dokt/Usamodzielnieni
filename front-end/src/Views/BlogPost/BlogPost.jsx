import React, { useState, useEffect, useContext } from "react";
import { Container, Card, Alert } from "react-bootstrap";
import { UserContext } from "context";

import {
  CommentForm,
  CommentsList,
  BlogContent
} from './_components';
import {getPost} from "./functions/apiCalls";



const mapPost = (res) => ({
  title: res.title,
  id: res.id,
  content: res.content,
  category: res.category,
  comments: res.comments.map((comment) => ({
    id: comment.id,
    content: comment.content,
    creationDate: comment.date_created,
    author: mapAuthor(comment.author)
  })),
  creationDate: res.date_created,
  author: mapAuthor(res.author),
  tags: res.tags
});

const mapAuthor = (author) => ({
  email: author.email,
  firstName: author.first_name,
  lastName: author.last_name
});

const BlogPost = () => {
  const [post, setPost] = useState(null);
  const [isPostLoading, setIsPostLoading] = useState(false);
  const [error, setError] = useState(false);
  const user = useContext(UserContext);

  const post_Id = window.location.pathname.replace(/\/blog\/blogpost\//, '');

  useEffect(
    () => {
      const loadPost = async (postId, token) => {
        setIsPostLoading(true);
        let loadedPost;
        try {
          loadedPost = mapPost(await getPost(postId, token));
        } catch (e) {
          console.log(e);
          loadedPost = null;
          setError(true);
        }
        setPost(loadedPost);
        setIsPostLoading(false);
      };
      loadPost(post_Id, user.token)
    },
    [post_Id, user.token]
  );

  const msg = error ? (<Alert variant="danger">Wystąpił błąd podczas wczytywania zawartości bloga.</Alert>) :
              isPostLoading ? (<Alert variant="info">Ładowanie zawartości bloga...</Alert>) :
              !post && (<Alert>null</Alert>);

  return msg || (
    <Container className="blogpost_container">
      <BlogContent post={post} type={user}/>
      <Card className="blogpost_comment_card">
        <Card.Body>
          <Card.Title as="h3" className="mb-3">Komentarze:</Card.Title>
          <CommentsList comments={post.comments} />
          <CommentForm />
        </Card.Body>
      </Card>
    </Container>
  );
};

export default BlogPost;
