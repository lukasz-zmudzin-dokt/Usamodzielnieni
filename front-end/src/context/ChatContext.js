import React, {
  createContext,
  useEffect,
  useState,
  useContext,
  useCallback,
} from "react";
import proxy from "config/api";
import { UserContext } from "context";

const getChats = async (token, filters) => {
  let url = `${proxy.chat}?page=${filters.page}&page_size=${filters.pageSize}`;
  const headers = {
    Authorization: "Token " + token,
    "Content-Type": "application/json",
  };

  const response = await fetch(url, { method: "GET", headers });

  if (response.status === 200) {
    return response.json().then((chats) => chats);
  } else {
    throw response.status;
  }
};

export const ChatContext = createContext(null);

export const ChatProvider = (props) => {
  const user = useContext(UserContext);
  const [chats, setChats] = useState([]);
  const [isChatsLoading, setIsChatsLoading] = useState(false);
  const [error, setError] = useState(false);
  const [count, setCount] = useState(0);
  const [filters] = useState({
    page: 1,
    pageSize: 10,
  });
  const [activePage, setActivePage] = useState(1);

  useEffect(() => {
    if (user.token) {
      const loadChats = async (token) => {
        setActivePage(1);
        setError(false);
        setIsChatsLoading(true);
        let loadedChats;
        try {
          loadedChats = await getChats(token, filters);
        } catch (e) {
          console.log(e);
          loadedChats = { count: 0, results: [] };
          setError(true);
        }
        setChats(loadedChats.results);
        setCount(loadedChats.count);
        setIsChatsLoading(false);
      };
      loadChats(user.token);
    }
  }, [filters, user.token]);

  const loadMoreMessages = async () => {
    let res;
    try {
      res = await getChats(user.token, { ...filters, page: activePage + 1 });
      setActivePage(activePage + 1);
    } catch (e) {
      setError(true);
      return;
    }

    setChats([...chats, ...res.results]);
  };

  const loadMessages = async (token) => {
    setActivePage(1);
    let res;
    try {
      res = await getChats(token, filters);
    } catch (e) {
      setError(true);
      return;
    }

    setChats(res.results);
  };

  const data = {
    chats,
    error,
    count,
    loadMoreMessages,
    isChatsLoading,
    loadMessages,
  };

  return <ChatContext.Provider value={data} {...props} />;
};
