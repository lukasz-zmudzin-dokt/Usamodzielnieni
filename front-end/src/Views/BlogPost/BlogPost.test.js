import React from "react";
import BlogPost from "./BlogPost";
import {render, waitForElement} from "@testing-library/react";
import {UserContext,AlertContext} from "context";


describe('BlogPost', () => {
    let post;
    let apiStatus;
    let user;
    let contextA = {
        open: true,
        changeVisibility: jest.fn(),
        message: "abc",
        changeMessage: jest.fn(),
    };
    beforeAll(() => {
        global.fetch = jest.fn().mockImplementation( (input, init) => {
            return new Promise((resolve, reject) => {
                switch (init.method) {
                    case "GET":
                        resolve ( apiStatus !== 200 ? {status: apiStatus} : {
                            status: 200,
                            json: () => Promise.resolve(post)
                        });
                        break;
                    default:
                        reject({});
                        break;
                }
            });
        });
        user = {
            type: "Staff",
            data: {
                email: "qwe@qwe.fgh"
            }
        }
    });

    beforeEach(() => {
        apiStatus = 200;
        post = {
            id: 123,
            author: {
                first_name: "Jan",
                last_name: "Kowalski",
                email: "jan@kowalski.pl"
            },
            content: "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Proin nibh augue, suscipit a, scelerisque sed, lacinia in, mi. Cras vel lorem. Etiam pellentesque aliquet tellus. Phasellus pharetra nulla ac diam. Quisque semper justo at risus. Donec venenatis, turpis vel hendrerit interdum, dui ligula ultricies purus, sed posuere libero dui id orci. Nam congue, pede vitae dapibus aliquet, elit magna vulputate arcu, vel tempus metus leo non est. Etiam sit amet lectus quis est congue mollis. Phasellus congue lacus eget neque. Phasellus ornare, ante vitae consectetuer consequat, purus sapien ultricies dolor, et mollis pede metus eget nisi. Praesent sodales velit quis augue. Cras suscipit, urna at aliquam rhoncus, urna quam viverra nisi, in interdum massa nibh nec erat.",
            date_created: "06-07-2019",
            category: "qwe",
            tags: ["tag1", "tag2", "tag3"],
            comments: [{
                author: {
                    first_name: "Jan",
                    last_name: "Nowak",
                    email: "qwe@qwe.qwe"
                }
            }]
        };
        jest.clearAllMocks();
    });

    it('should match snapshot after loaded', async () => {
        const { container, getByText } = render(
            <UserContext.Provider value={user}>
                <BlogPost/>
            </UserContext.Provider>
        );
        await waitForElement(() => getByText('Lorem ipsum dolor', {exact: false}));

        expect(container).toMatchSnapshot();
    });

    it('should render loading when loading lol', async() => {
        const {getByText, queryByText} = render(
            <UserContext.Provider value={user}>
                <BlogPost/>
            </UserContext.Provider>
        );

        expect(getByText("Ładowanie", {exact: false})).toBeInTheDocument();
        expect(queryByText("Lorem ipsum dolor", {exact: false})).not.toBeInTheDocument();
        await waitForElement(() => getByText("Lorem ipsum dolor", {exact: false}));
    });

    it('should render error on api fail', async() => {
        apiStatus = 500;
        const {getByText, queryByText} = render(
            <UserContext.Provider value={user}>
                <AlertContext.Provider value={contextA}>
                    <BlogPost/>
                </AlertContext.Provider>
            </UserContext.Provider>
        );

        await waitForElement(() => getByText("null", {exact: false}));

        expect(contextA.changeMessage).toHaveBeenCalledWith(
          "Wystąpił błąd podczas wczytywania zawartości bloga."
        );

        expect(queryByText("Lorem ipsum dolor", {exact: false})).not.toBeInTheDocument();
    });

    it('should render comments', async () => {
        const {getByText} = render(
            <UserContext.Provider value={user}>
                <BlogPost/>
            </UserContext.Provider>
        );

        await waitForElement(() => fetch);
        expect(getByText("Jan Nowak", {exact: false})).toBeInTheDocument();
    });
});