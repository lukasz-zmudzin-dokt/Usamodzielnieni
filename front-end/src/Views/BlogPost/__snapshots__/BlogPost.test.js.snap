// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlogPost should match snapshot after loaded 1`] = `
<div>
  <div
    class="blogpost_container container"
  >
    <div
      class="card"
    >
      <div
        class="card-header"
      />
      <div
        class="post_content mx-4 card-body"
      >
        <h3
          class="post_title card-title"
        >
          Tytuł posta
        </h3>
        <h6
          class="text-muted mb-4 mt-2 card-subtitle"
        >
          Kategoria: 
          qwe
        </h6>
        <p
          class="blog_content_text text-justify card-text"
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Proin nibh augue, suscipit a, scelerisque sed, lacinia in, mi. Cras vel lorem. Etiam pellentesque aliquet tellus. Phasellus pharetra nulla ac diam. Quisque semper justo at risus. Donec venenatis, turpis vel hendrerit interdum, dui ligula ultricies purus, sed posuere libero dui id orci. Nam congue, pede vitae dapibus aliquet, elit magna vulputate arcu, vel tempus metus leo non est. Etiam sit amet lectus quis est congue mollis. Phasellus congue lacus eget neque. Phasellus ornare, ante vitae consectetuer consequat, purus sapien ultricies dolor, et mollis pede metus eget nisi. Praesent sodales velit quis augue. Cras suscipit, urna at aliquam rhoncus, urna quam viverra nisi, in interdum massa nibh nec erat.
        </p>
        <p
          class="post_taglist mt-5"
        >
          <em>
            tagi: 
            <span
              class="mx-1 badge badge-info"
            >
              tag1
            </span>
            <span
              class="mx-1 badge badge-info"
            >
              tag2
            </span>
            <span
              class="mx-1 badge badge-info"
            >
              tag3
            </span>
          </em>
        </p>
      </div>
      <div
        class="blogpost_summary card-footer"
      >
        <div
          class="row"
        >
          <div
            class="mx-3 col"
          >
            <div
              class="row"
            >
              Autor: Jan Kowalski (jan@kowalski.pl)
            </div>
            <div
              class="row"
            >
              Opublikowano: 06.07.2019
            </div>
          </div>
          <div
            class="post_comment_counter"
          >
            <div
              class="text-right mx-3 row"
            >
              Liczba komentarzy: 
              0
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="blogpost_comment_card card"
    >
      <div
        class="card-body"
      >
        <h3
          class="mb-3 card-title"
        >
          Komentarze:
        </h3>
        <div
          class="fade alert alert-info show"
          role="alert"
        >
          Brak komentarzy.
        </div>
        <div
          class="card"
        >
          <div
            class="card-body"
          >
            <div
              class="card-title h5"
            >
              Dodaj komentarz
            </div>
            <form
              class=""
            >
              <div
                class="form-group"
              >
                <textarea
                  class="form-control"
                  id="commentContent"
                  placeholder="Treść komentarza"
                  required=""
                />
              </div>
              <div
                class="CommentForm__submitGroup mb-0 form-group"
              >
                <button
                  class="btn btn-primary"
                  type="submit"
                >
                  Prześlij
                </button>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
